src/View/Sidebar/SidebarWindow.vala

```diff
@@ -0,0 +1,289 @@
+            orientation = Gtk.Orientation.VERTICAL,
+            vexpand = true
+        };
+        bookmarklists_grid.add (bookmark_expander);
+        bookmarklists_grid.add (bookmark_revealer);
+        bookmarklists_grid.add (device_expander);
+        bookmarklists_grid.add (device_revealer);
+        bookmarklists_grid.add (network_expander);
+        bookmarklists_grid.add (network_revealer);
+
+        scrolled_window = new Gtk.ScrolledWindow (null, null);
+        scrolled_window.add (bookmarklists_grid);
+
+        var connect_server_button = new Gtk.Button.with_label (_("Connect Server…")) {
+            always_show_image = true,
+            image = new Gtk.Image.from_icon_name ("network-server-symbolic", Gtk.IconSize.MENU),
+            tooltip_markup = Granite.markup_accel_tooltip ({"<Alt>C"})
+        };
+
+        var action_bar = new Gtk.ActionBar ();
+        action_bar.get_style_context ().add_class (Gtk.STYLE_CLASS_FLAT);
+        action_bar.add (connect_server_button);
+
+        orientation = Gtk.Orientation.VERTICAL;
+        width_request = Marlin.app_settings.get_int ("minimum-sidebar-width");
+        get_style_context ().add_class (Gtk.STYLE_CLASS_SIDEBAR);
+        add (scrolled_window);
+        add (action_bar);
+
+        plugins.sidebar_loaded (this);
+
+        reload ();
+        show_all ();
+
+        Marlin.app_settings.bind (
+            "sidebar-cat-personal-expander", bookmark_expander, "active", SettingsBindFlags.DEFAULT
+        );
+        Marlin.app_settings.bind (
+            "sidebar-cat-devices-expander", device_expander, "active", SettingsBindFlags.DEFAULT
+        );
+        Marlin.app_settings.bind (
+            "sidebar-cat-network-expander", network_expander, "active", SettingsBindFlags.DEFAULT
+        );
+
+        bookmark_expander.bind_property ("active", bookmark_revealer, "reveal-child", GLib.BindingFlags.SYNC_CREATE);
+        device_expander.bind_property ("active", device_revealer, "reveal-child", GLib.BindingFlags.SYNC_CREATE);
+        network_expander.bind_property ("active", network_revealer, "reveal-child", GLib.BindingFlags.SYNC_CREATE);
+
+        connect_server_button.clicked.connect (() => {
+            connect_server_request ();
+        });
+    }
+
+    private void refresh (bool bookmarks = true, bool devices = true, bool network = true) {
+        //Do not refresh if already refreshing or if will be reloaded anyway
+        if (loading || reload_timeout_id > 0) {
+            return;
+        }
+
+        loading = true;
+
+        if (bookmarks) {
+            bookmark_listbox.refresh ();
+        }
+
+        if (devices) {
+            device_listbox.refresh ();
+        }
+
+        if (network) {
+            network_listbox.refresh ();
+        }
+
+        loading = false;
+    }
+
+    /* SidebarInterface */
+    public uint32 add_plugin_item (Marlin.SidebarPluginItem plugin_item, Marlin.PlaceType category) {
+        uint32 id = -1;
```

Comment:
I don't really like returning -1 with uint, can we find a better value like 0 or uint32.MAX (even if -1 is the same, that's more visible that way)

Reply:
I think we can use 0.  Real items have an id > 0.